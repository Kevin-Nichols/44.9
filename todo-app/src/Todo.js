import React, { useState } from "react";

function Todo(
  {todo="default thing todo", 
  id="1", 
  remove, 
  update}
  ) {
  const [editTodo, setEditTodo] = useState(todo);
  const [editing, setEditing] = useState(false);
  
  const handleChange = e => {
    setEditTodo(e.target.value);
  };

  const editOnOff = () => {
    setEditing(edit => !edit);
  };

  const handleRemove = () => remove(id);

  const handleUpdate = e => {
    e.preventDefault();
    update(id, editTodo);
    setEditing(false);
  };

  //Normale JSX
  let todoJsx = (
    <>
      <li>{todo}</li>
      <button onClick={editOnOff}>Edit Todo</button>
      <button onClick={handleRemove}>X</button>
    </>
  );

  //JSX if editing
  if(editing) {
    todoJsx = (
      <>
        <form onSubmit={handleUpdate}>
          <input type="text" onChange={handleChange} value={editTodo} />
          <button>Update Todo</button>
        </form>
      </>
    );
  };

  return todoJsx;
}

export default Todo;