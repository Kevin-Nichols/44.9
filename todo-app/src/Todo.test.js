import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Todo from "./Todo";

it("is a smoke test for Todo", function() {
  render(<Todo />);
});

it("is a snapshot test for Todo", function() {
  const { asFragment } = render(<Todo />);
  expect(asFragment()).toMatchSnapshot();
});

it("is a snapshot test when editing", function() {
  const { getByText, asFragment } = render(<Todo />);
  const editButton = getByText("Edit Todo");
  fireEvent.click(editButton);
  expect(asFragment()).toMatchSnapshot();
});

it("updates on form submission", function() {
  const update = jest.fn();
  const { getByText } = render(<Todo update={update} />);
  const editButton = getByText("Edit Todo");
  fireEvent.click(editButton);
  const updateButton = getByText("Update Todo");
  fireEvent.click(updateButton);
  expect(update).toHaveBeenCalled();
});

it("deletes on button click", function() {
  const remove = jest.fn();
  const { getByText } = render(<Todo remove={remove} />);
  const deleteButton = getByText("X");
  fireEvent.click(deleteButton);
  expect(remove).toHaveBeenCalled();
});