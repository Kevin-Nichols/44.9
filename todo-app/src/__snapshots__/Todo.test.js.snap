// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is a snapshot test for Todo 1`] = `
<DocumentFragment>
  <li>
    default thing todo
  </li>
  <button>
    Edit Todo
  </button>
  <button>
    X
  </button>
</DocumentFragment>
`;

exports[`is a snapshot test when editing 1`] = `
<DocumentFragment>
  <form>
    <input
      type="text"
      value="default thing todo"
    />
    <button>
      Update Todo
    </button>
  </form>
</DocumentFragment>
`;
